### ISSUE TITLE
Volcabulary, Marketing & Organizational Ideas

### ISSUE THREAD

**Issue by matthewmueller** [makojs/core#24]
I've been thinking about how to get this library "production-ready" and to be promoted and shared. I wanted to propose a few changes.

#### Rename analyze and assemble phases

Right now I find this confusing because the analyze phase builds up the dependency graph and the assemble phase breaks it down. Assemble to me means, "create", or build up the tree.

I think we should go with:

**analyze → parse**
**assemble → compile (or generate)**

Parse is familiar to people and typically means take in a string and spit out an AST (or tokens), which is not unlike the current analyze phase. Next thing we do is take this tree, rejigger it, collapse it and then use it to make something useful. I think this could be either compiling or generating.

#### Rename tree to graph

This one I don't have a huge opinion on, but it's not really a tree we're operating on, it's a directed graph.

#### Hooks

The hooks (and their corresponding explanations in the wiki) are probably the most powerful part of mako. 

I'm not a huge fan of the new `preanalyze`, `postanalyze`, `preassemble` and `postassemble` hooks. I think they were added to support bundles, but that could be added with one new hook. I think hooks should be added on a case by case basis, not for consistency purposes. I'd like to propose some changes:

I. Analyze phase:

  **preread → read → postread → predependencies → dependencies**

- just remove preanalyze and postanalyze

II. Assemble phase:

  **prebundle → bundle → prewrite → write → postwrite**
  - rename postdependencies to bundle. i found this quite unclear because it seemed like postdependencies was an extension of the analyze phase, but it's really a part of the assemble phase.
  - add prebundle. prebundle should cover the common bundling stuff, it can also be where treeshaking happens, getting/generating CDN resource URLs, and other tree-level optimizations.

#### Introduce the concept of presets (as a bundle of configured plugins)

This idea is borrowed from babel and is what I wanted to do with scooby. Basically create configuration- free bundles of plugins, where I can require `mako-preset-js`, `mako-preset-jade`, include them in the `package.json` and voila, you have a working build system.

There are actually a ton of right and wrong decisions with your build system and presets would help newcomers get up and running with good defaults. You wouldn't need to think about how to do sourcemaps in production, minification, etc. Minimal configuration can be done via environment variables (like inserting S3 credentials).

#### Rename mako

This is probably the most controversial, but I'd like to propose a name change. This is totally my opinion but I don't like the name for 2 reasons:

- It sounds like an off-brand of make. Like many of the other make flavors that have come and went (cake , jake, etc). This idea fundamentally different than those make flavors
- It also sounds sort of masculine and macho. Ex. doesn't "maka" sound more feminine? I'd prefer something a little more fun to say and neutral

I can offer up 3 names: *thimble*, *nom* & *duo*.

I prefer thimble the most because you get a cute logo (ex. http://matthewmueller.github.io/thimble/thimble.png) and it's easy to come up with a tagline (Elegant asset pipeline for Makers).

For duo, I don't think people would mind if we just bump major and call it a day. Not too many people use it anymore. Duo is short and it reminds me of having my asset pipeline being my trusty friend, which is very much true haha.

#### Flow Diagram

Would be nice to have some sort of diagram illustrating the flow (alongside the wiki explanations of each hook). It'd be cool if it was interactive but that might be for later. Something like this: https://cldup.com/Sc8FHxzYOE.jpeg

#### Monorepo 👯 

I started playing with mako as a monorepo and I'm now convinced for big projects with lots of plugins that change quickly like this one, monorepo (with proper tooling) is the way to go. [lerna](https://github.com/lerna/lerna) gives your monorepo the same benefits of breaking everything up (like independently installing, testing & publishing separately). I can talk about this more if you'd like to consider it or set it up as a separate branch.

#### Use standard

I think this will encourage more PRs in the future. I think the community has sort of converged on just using standard. When I started using it, I was a little annoyed that it wasn't my perfect syntax, but it's fine now. Whenever I contribute to mako, I have to remap my thinking to your custom rules, which aren't bad, just more work on my end.

---

Excited to see where this heads! I'm convinced this is the way forward out of the crazy webpack mess.

**Comment by dominicbarnes** [comment]
@matthewmueller I'm really digging a lot of what I've read so far. It's pretty late for me, so I'm going to sleep on all of this, but I will give a thorough response to everything this weekend.

It means a lot to have others like yourself really thinking hard about this project, and about how best to move forward. Thanks for taking the time! (you'll be hearing from me again soon enough)

**Comment by matthewmueller** [comment]
Thanks man, of course – take your time. Lots of personal taste & opinions here too. Looking forward to getting some feedback.

**Comment by dominicbarnes** [comment]
First, in regards to renaming "analyze" and "assemble":

> Parse is familiar to people and typically means take in a string and spit out an AST (or tokens), which is not unlike the current analyze phase.

I don't think the word "parse" really fits here. Lots of plugins for this phase are parsers, but not mako itself. I prefer "analyze" in this case. (but I'm still open to other options)

> Next thing we do is take this tree, rejigger it, collapse it and then use it to make something useful. I think this could be either compiling or generating.

Totally agree, I really like using the word "compile" here.

**Comment by dominicbarnes** [comment]
Next, let's talk about the hooks:

> I'm not a huge fan of the new preanalyze, postanalyze, preassemble and postassemble hooks. I think they were added to support bundles, but that could be added with one new hook. I think hooks should be added on a case by case basis, not for consistency purposes.

I disagree, I think the build-level hooks were a natural extension of the other hooks. Yeah, I initially added it for bundle support, and for that alone I find them very useful and worthwhile. Not to mention, my next iteration of mako-serve and mako-cli will leverage all the hooks when writing human-friendly output. (instead of the awkward event emitter I have atm)

> rename postdependencies to bundle. i found this quite unclear because it seemed like postdependencies was an extension of the analyze phase, but it's really a part of the assemble phase.

Yeah, I see what you mean, I always felt like `postdependencies` was weirdly named. However, I do prefer the consistency of the naming now, and I think documentation can help with this 1 oddity. I personally find replacing with `prebundle` and `bundle` to be way too specific for my taste.

**tl;dr:** I want to leave the hooks as-is for now.

**Comment by dominicbarnes** [comment]
As for the name "mako", here's where it came from:

 - [a specific species of shark](https://en.wikipedia.org/wiki/Mako_shark_(disambiguation)) known for being the [fastest](https://en.wikipedia.org/wiki/Shortfin_mako_shark#Behavior) (sharks are my favorite animals)
 - [a core component to the lifeforce of Gaia](http://finalfantasy.wikia.com/wiki/Mako) in Final Fantasy VII, it acts as both an energy source and enabling the use of many special abilities.

The name is fairly arbitrary, but I'm still fond of it. (as you can tell from my Twitter handle "mako281")

> It sounds like an off-brand of make. Like many of the other make flavors that have come and went (cake , jake, etc). This idea fundamentally different than those make flavors

I see where you're coming from, and it's compelling to me because I don't want this tool to get dismissed or lumped in with others that easily.

> It also sounds sort of masculine and macho. Ex. doesn't "maka" sound more feminine? I'd prefer something a little more fun to say and neutral

I never really thought of it as masculine/macho, maybe it depends on how you pronounce it? I say "may-koh", as opposed to "mah-koh", not sure if that even makes a difference.

As far as your 3 suggestions:

 > thimble

It's a nice name, easily my favorite of the 3. I do like the idea of tying it with "makers", and getting a simple logo is a great plus.

 > nom

This one is just weird to me, and you didn't really give any further context/explanation, so I'm just gonna veto this one now.

 > duo

I thought about this long and hard. On one hand, being able to tap into the remains of the duo community is appealing compared to starting from scratch. On the other hand, I really feel like this is a completely different tool in practically every way, so to assume duo's identity doesn't sit right with me.

All that said, while I'm not entirely against a name change, I'm not especially excited about these 3 options. I still like the name "mako", and maybe adding a bit more context from my end makes it a bit more appealing to you.

**Comment by dominicbarnes** [comment]
> Would be nice to have some sort of diagram illustrating the flow (alongside the wiki explanations of each hook)

I **really** like this idea, it would really help crystallize all these crucial concepts for new users/developers.

**Comment by dominicbarnes** [comment]
> Monorepo

I'm not a fan of monorepos, and I personally find it really hard to contribute on them, as they're big/complex by definition. (my prime example here is babel's repo)

I think the mental model of individual repos is much preferable to a single monolith. That said, I'm going to decline this idea. (may reconsider in the future)

**Comment by dominicbarnes** [comment]
As far as adopting standard, I think you have a lot of great points. While I do prefer my own style, I think community involvement is a much better goal of mine.

**Comment by dominicbarnes** [comment]
@matthewmueller sorry to talk your ear off lol.

I'm done adding my initial feedback, and I look forward to hearing back from you!

**Comment by dominicbarnes** [comment]
Oh yeah, almost forgot about renaming tree to graph. I disagree, but only on a technicality. True, you can add cycles to the graph during "analyze", which does contradict the name tree. However, the first step in building removes cycles making it back into a tree again.

Also, I don't want to use `graph.graph`, which would happen if I called my abstraction a graph, and then added a graph.js instance to it.

**Comment by matthewmueller** [comment]
Cool, sounds good. And thanks for the in-depth response. Here's a few additional thoughts and clarification:

> I don't think the word "parse" really fits here. Lots of plugins for this phase are parsers, but not mako itself. I prefer "analyze" in this case. (but I'm still open to other options)

I realized I didn't finish the argument, not sure it makes much a difference but what i meant by `parse()` turning strings into an AST, is that in the same way that you create an abstract syntax tree representation of a language, you are creating a dependency tree of your project.

>  I personally find replacing with prebundle and bundle to be way too specific for my taste.

Bundle might be an overloaded term, but I thought of bundle more in terms of a synonym of **reduction**, which is what we're doing – reducing and extracting values from the dependency tree.

> As for the name "mako", here's where it came from:

Okay cool, I think a nice logo would help clarify this. If I get some time, I'll throw together some options.

> I disagree, I think the build-level hooks were a natural extension of the other hooks. Yeah, I initially added it for bundle support, and for that alone I find them very useful and worthwhile.

My only concern with this is that adding too many hooks gets confusing on where you should be extending mako. Documentation can help with this though.

Also I'm assuming your going to be using these hooks for logging in serve and CLI. These hooks aren't quite powerful enough for that. I think that logging should be enabled on a per-file, per-stage basis, so I should be able to hook into logging this file at this stage in the system. This would allow a CLI that lets you do things like this:

**before**

```
index.js → preread → ...working...
index.css → preread → read → predependencies → ...working...
```

**some time later**

```
index.js → preread → read → ...working...
index.css → preread → read → predependencies → dependencies → ...working...
```

**Comment by matthewmueller** [comment]
Random note, but I've been thinking about it. I think in the documentation, we should have names for the two different trees produced. I propose:

- *dependency tree* for the tree that comes out after the `analyze` phase
- *build tree* for the tree that comes out after the `assemble` phase

**Comment by dominicbarnes** [comment]
I _really_ like that idea, that makes it easier to understand how those phases differ.

**Comment by dominicbarnes** [comment]
@matthewmueller: I remembered one other thing, regarding presets. This is already supported and used by mako, anything passed to `.use()` can either call `.use()` itself or be an array of plugins. mako-browser is the prime example of this concept, it's my "preset" for a component/duo/browserify/webpack-style workflow.

**Comment by darsain** [comment]
> mako sounds sort of masculine and macho

This has to be affected by your first language, because for me it sounds like a name for a teddy bear :)

Regarding standard, I'd love to evangelize [XO](https://github.com/sindresorhus/xo), but I don't *really* care :)

I have nothing else to contribute here :D

**Comment by dominicbarnes** [comment]
I'm gonna start splitting off individual tasks here into separate issues. Once I've done that, I'll probably close this issue itself.

**Comment by dominicbarnes** [comment]
Thanks for taking all the time to write this up @matthewmueller, and thanks @darsain for your contributions here as well. Definitely looking forward to what the future holds :)

**Comment by matthewmueller** [comment]
❤️ thanks for all the hard work!
